app/api/upstash/[action]/[...key]/route.ts

```diff
@@ -2,13 +2,42 @@ import { NextRequest, NextResponse } from "next/server";
 
 async function handle(
   req: NextRequest,
-  { params }: { params: { path: string[] } },
+  { params }: { params: { action: string; key: string[] } },
 ) {
+  const requestUrl = new URL(req.url);
+  const endpoint = requestUrl.searchParams.get("endpoint");
+
   if (req.method === "OPTIONS") {
     return NextResponse.json({ body: "OK" }, { status: 200 });
   }
+  const [action, ...key] = params.key;
+  // only allow to request to *.upstash.io
+  if (!endpoint || !endpoint.endsWith("upstash.io")) {
```

Comment:
is string `endsWith` enough here? It seems still possible to craft a endpoint like `https://arbitrary.url.i.want/path?dummy=upstash.io` and bypass.

What about 

```diff
try {
  if ( ! (new URL(endpoint).hostname.endswith('.upstash.io')   )  {
   throw 403
 }
} catch (e) {
  throw 400
}

Reply:
Thanks for reviewing the code. Just updated with the new matching method.